faculty/clients/serveragent.py

```diff
@@ -0,0 +1,344 @@
+        current cache memory usage.
+    rss: float
+        resident set size allocation. 
+    """
+
+    total = attrib()
+    used = attrib()
+    cache = attrib()
+    rss = attrib()
+
+
+class ServerAgentClient(BaseClient):
+    """Client for the Faculty server events service.
+
+    Build this client with a session directly.
+
+    >>> client = ServerAgenClient(url, session)
+
+    Parameters
+    ----------
+    url : str
+        The URL of the object storage service.
+    session : faculty.session.Session
+        The session to use to make requests.
+    """
+    def latest_environment_execution(self):
+        """Get the latest environment execution on the server.
+        
+        Returns
+        -------
+        Execution
+            The latest environment execution on a server. 
+        """
+        return self._get("/execution/latest", _ExecutionSchema())
+
+    def stream_server_events(self, endpoint):
+        """Read the server events stream from an endpoint.
+        
+        Parameters
+        ----------
+        endpoint : str
+            HTTP request endpoint.
+
+        Yields
+        ------
+        ServerSentEventMessage
+        """
+        with self._stream(endpoint) as stream:
+            for message in stream:
+                yield message
+
+    def stream_server_resources(self):
+        """Stream the resources used by the server.
+        
+        Yields
+        ------
+        ServerResources
+        """
+
+        schema = _ServerResourcesSchema()
+        for message in self.stream_server_events("/events"):
+            if message.event == SERVER_RESOURCES_EVENT:
+                yield schema.load(json.loads(message.data))
+
+    def stream_environment_execution_step_logs(self, execution_id, step_id):
+        """Read from the environment step logs.
+        
+        Parameters
+        ----------
+        execution_id : Execution.id
+            ID of the environment execution.
+        step_id : EnvironmentExecutionStep.id
+            ID of the environment execution step.
+
+        Yields
+        ------
+        EnvironmentExecutionStepLog
+
+        """
+        endpoint = "/execution/{}/executor/{}/logs".format(
+            execution_id, step_id
+        )
+        schema = _EnvironmentExecutionStepLogSchema()
+        for message in self.stream_server_events(endpoint):
+            if message.event == "log":
+                for line in json.loads(message.data):
+                    yield schema.load(line)
+
+    
+
+class _EnvironmentExecutionStepLogSchema(BaseSchema):
+
+    line_number = fields.Integer(data_key="lineNumber", required=True)
+    content = fields.String(required=True)
+
+    @post_load
+    def make_environment_execution_step_log(self, data, **kwargs):
+        return EnvironmentExecutionStepLog(**data)
+
+
+class _EnvironmentExecutionStepSchema(BaseSchema):
+
+    id = fields.UUID(required=True)
+    command = fields.List(fields.String(required=True), required=True)
```

Comment:
`required=True` is redundant for entries of a list (if an entry is missing, it'll mean there's an empty list, which is valid).

```suggestion
    command = fields.List(fields.String(), required=True)
```